src/PowerShellEditorServices/Services/Analysis/PssaCmdletAnalysisEngine.cs

```diff
@@ -0,0 +1,544 @@
+
+            IReadOnlyCollection<PSObject> diagnosticResults = result?.Output ?? s_emptyDiagnosticResult;
+            _logger.LogDebug(String.Format("Found {0} violations", diagnosticResults.Count));
+
+            var scriptMarkers = new ScriptFileMarker[diagnosticResults.Count];
+            int i = 0;
+            foreach (PSObject diagnostic in diagnosticResults)
+            {
+                scriptMarkers[i] = ScriptFileMarker.FromDiagnosticRecord(diagnostic);
+                i++;
+            }
+
+            return scriptMarkers;
+        }
+
+        private Task<PowerShellResult> InvokePowerShellAsync(PSCommand command)
+        {
+            return Task.Run(() => InvokePowerShell(command));
+        }
+
+        private PowerShellResult InvokePowerShell(PSCommand command)
+        {
+            using (var powerShell = System.Management.Automation.PowerShell.Create())
+            {
+                powerShell.RunspacePool = _analysisRunspacePool;
+                powerShell.Commands = command;
+                PowerShellResult result = null;
+                try
+                {
+                    Collection<PSObject> output = InvokePowerShellWithModulePathPreservation(powerShell);
+                    PSDataCollection<ErrorRecord> errors = powerShell.Streams.Error;
+                    result = new PowerShellResult(output, errors, powerShell.HadErrors);
+                }
+                catch (CommandNotFoundException ex)
+                {
+                    // This exception is possible if the module path loaded
+                    // is wrong even though PSScriptAnalyzer is available as a module
+                    _logger.LogError(ex.Message);
+                }
+                catch (CmdletInvocationException ex)
+                {
+                    // We do not want to crash EditorServices for exceptions caused by cmdlet invocation.
+                    // Two main reasons that cause the exception are:
+                    // * PSCmdlet.WriteOutput being called from another thread than Begin/Process
+                    // * CompositionContainer.ComposeParts complaining that "...Only one batch can be composed at a time"
+                    _logger.LogError(ex.Message);
+                }
+
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// Execute PSScriptAnalyzer cmdlets in PowerShell while preserving the PSModulePath.
+        /// Attempts to prevent PSModulePath mutation by runspace creation within the PSScriptAnalyzer module.
+        /// </summary>
+        /// <param name="powershell">The PowerShell instance to execute.</param>
+        /// <returns>The output of PowerShell execution.</returns>
+        private Collection<PSObject> InvokePowerShellWithModulePathPreservation(System.Management.Automation.PowerShell powershell)
+        {
+            if (_alreadyRun)
```

Comment:
What is this check for?

Reply:
Since runspace creation only happens at first invocation, I put a boolean in to reduce the overhead on subsequent calls

Reply:
Run a few manual tests to make sure - I know RunspacePool has a habit of creating new runspaces within the pool (as in, we don't have control over when, why and how it creates Runspaces). I've seen in `Get-Runspace` incremental runspace numbering which is why I have concerns here.